### ISSUE TITLE
Security weight greater than 1.0

### ISSUE THREAD

**Issue by sven337** [pmorissette/bt#80]
This may be similar to #67.
I was playing with get_transactions recently and observed something strange: there are spurious transactions for tiny amounts.
I'm testing a momentum strategy that selects the best one performing asset and allocates everything to it. (SelectN(1) then WeighEqually(), which I've verified does set a weight of exactly 1.0 for the desired security).

So the expectation is that every month, either nothing at all will happen (keeping the same security), or the existing position will be completely closed out and capital allocated to another security.
In other words, security weights are supposed to be 0 everywhere and exactly 1 elsewhere.

However, here are the weights reported by res.backtest['s1'].security_weights['GRE.PA']:
```
2012-10-01    1.008000
2012-10-02    1.007680
2012-10-03    1.007704
2012-10-04    1.007593
[...]
2012-11-01    1.001274
```

That's obviously impossible.

Here are the transactions:
```
2012-10-01 GRE.PA      1.260  617.071981
           LVE.PA     12.215  -63.815857
2012-11-01 GRE.PA      1.284   -4.805857
2012-12-03 GRE.PA      1.341   -0.745712
2013-01-01 GRE.PA      1.455   -0.687285
2013-02-01 GRE.PA      1.546   -0.646831
2013-03-01 GRE.PA      1.522 -610.186297
           THA.PA    158.880    5.800615
```
Note how GRE.PA is purchased after closing out LVE.PA. Then, for the following months until 2013-03-01, the strategy  has elected to stay on GRE.PA, so nothing should happen at all. Note that there are small sale transactions being reported! Any idea why that is?

I've verified that the algo does really ask for a weight of 1.0, so the problem isn't coming from there. Also, 1.008000 vs. 1.00000 isn't a believable floating point precision issue, at least not without a bug in bt. 
I've looked a little but haven't managed to figure out where the problem might come from. Your help would be appreciated!
Thanks!

**Comment by pmorissette** [comment]
Hey @sven337,

Thanks for reporting this. I will look into it and report back as soon as possible.

Thanks,
Phil

**Comment by pmorissette** [comment]
Hey @sven337,

So I dug a bit deeper. First, it's hard to know exactly what is going on in your case. If you could provide me with a simple example that exhibits this behavior, I could debug the problem. Also, are you running `bt` off a pip install, or are you running from `master` on github?

I was running some tests on my end and I did notice some strange trades. Here is some sample code:

```python
import bt
data = bt.get('aapl', start='2010-01-01')
s = bt.Strategy('s', [bt.algos.SelectAll(), bt.algos.WeighEqually(), bt.algos.Rebalance()])
t = bt.Backtest(s, data, commissions=lambda x, y: 0, integer_positions=False, initial_capital=1000001)
r = bt.run(t)
print r.get_transactions('s')
r.backtests['s'].security_weights['aapl'].max()
```

So basically you should only have 1 trade on day 1 and that's it. Interestingly, the output looks like this:

```
                         price      quantity
Date       Security                         
2010-01-04 aapl      27.727039  3.606588e+04
2010-01-05 aapl      27.774976 -7.275958e-12
2010-01-06 aapl      27.333178  7.275958e-12

1.0000000000000002
```

If you change initial_capital to 1000001, this issue disappears:

```
                         price      quantity
Date       Security                         
2010-01-04 aapl      27.727039  36065.913854

1.0
```

Therefore, I suspect there may indeed be some kind of rounding issue. That being said, I'm not sure this is the same issue you are dealing with (the weights seem too far off for rounding, as you mentioned), but with some sample code, I should be able to track it down. 

Thanks,
Phil

**Comment by sven337** [comment]
Hi @pmorissette,
thanks for looking into the problem. These spurious transactions you're listing above are an issue - they are what prompted my digging further. Using integer positions *seems* to be sufficient in making these transactions go away, but I'll double check.

As for the greater-than-1 weights, I tried to build a minimal repro case for you. Using your own case, if you don't set the "commissions" parameter, you'll observe the issue, even with integer positions:
```
                         price  quantity
Date       Security                     
2010-01-04 aapl      27.727039   36065.0
2010-01-05 aapl      27.774976     -13.0
Out[296]: 1.0003364328695146
```

In other words, the bug happens when you don't set a commission. I don't think this is expected?

Thanks!

**Comment by FinQuest** [comment]
Hi @sven337, looking at similar/same issue  (#82).  I think when you don't set the commissions parameter a default commission is used:

```python
# default commission function
def _dflt_comm_fn(self, q, p):
        return max(1, abs(q) * 0.01)
```

The following produce different transaction logs:

```python
bt_gross = bt.Backtest(strategy, data, integer_positions=False,
                       commissions=lambda q, p: 0, name='gross')                       
bt_no_comm = bt.Backtest(strategy, data, integer_positions=False,
                         name='no_comm')
```

**Comment by pmorissette** [comment]
@sven337, @FinQuest 

Working on this now. The initial assumption was that commissions were 'absorbed' by the capital in the strategy. In some cases, this meant that we actually had negative cash, which in turn led to these strange numbers. All-in-all, it wasn't an issue for me at the time and it simplified the code.

Anyways, I'm in the process of re-writing some of this logic and should have something working soon. This should eliminate these spurious trades and keep weights below 1. I just want to make sure I don't introduce another subtle bug by modifying this. 

Basically, the new strategy is to factor in the commission cost when trying to determine the number of shares, and reducing the number if need be. 

Will report back shortly.

**Comment by pmorissette** [comment]
Ok I just pushed up an update. You should now be able to upgrade through pip, or pull the latest master and run `make`.

Basically, I was previously ignoring commissions when calculating the quantity of stocks to buy/sell given an allocation. This was effectively like charging the commissions on some kind of margin account. 

Anyways, I updated the core logic in `SecurityBase.allocate` to factor in the cost of the commissions when determining the amount. In the event that the total outlay > allocation amount, the quantity will be decreased. 

In general, I think this will greatly reduce the spurious trades we were previously seeing. Of course, when using `integer_positions=False`, you will have some of these trades as the model attempts to rebalance towards the target weights. 

I hope this makes sense and let me know if you have any questions/issues. 

Cheers,
Phil

**Comment by sven337** [comment]
Thank you very much! 
I confirm that the update makes the issue that was bothering me go away.
